src/libraries/System.Text.RegularExpressions/tests/Regex.Count.Tests.cs

```diff
@@ -0,0 +1,120 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Threading.Tasks;
+using Xunit;
+
+namespace System.Text.RegularExpressions.Tests
+{
+    public class RegexCountTests
+    {
+        [Theory]
+        [MemberData(nameof(Count_ReturnsExpectedCount_TestData))]
+        public async Task Count_ReturnsExpectedCount(RegexEngine engine, string pattern, string input, RegexOptions options, int expectedCount)
+        {
+            Regex r = await RegexHelpers.GetRegexAsync(engine, pattern, options);
+            Assert.Equal(expectedCount, r.Count(input));
+            Assert.Equal(r.Count(input), r.Matches(input).Count);
+
+            if (options == RegexOptions.None && engine == RegexEngine.Interpreter)
+            {
+                Assert.Equal(expectedCount, Regex.Count(input, pattern));
+            }
+
+            switch (engine)
+            {
+                case RegexEngine.Interpreter:
+                case RegexEngine.Compiled:
+                case RegexEngine.NonBacktracking:
+                    RegexOptions engineOptions = RegexHelpers.OptionsFromEngine(engine);
+                    Assert.Equal(expectedCount, Regex.Count(input, pattern, options | engineOptions));
+                    Assert.Equal(expectedCount, Regex.Count(input, pattern, options | engineOptions, Regex.InfiniteMatchTimeout));
+                    break;
+            }
+        }
+
+        public static IEnumerable<object[]> Count_ReturnsExpectedCount_TestData()
+        {
+            foreach (RegexEngine engine in RegexHelpers.AvailableEngines)
+            {
+                yield return new object[] { engine, @"", "", RegexOptions.None, 1 };
+                yield return new object[] { engine, @"", "a", RegexOptions.None, 2 };
+                yield return new object[] { engine, @"", "ab", RegexOptions.None, 3 };
+
+                yield return new object[] { engine, @"\w", "", RegexOptions.None, 0 };
+                yield return new object[] { engine, @"\w", "a", RegexOptions.None, 1 };
+                yield return new object[] { engine, @"\w", "ab", RegexOptions.None, 2 };
```

Comment:
possibly a case with a loop eg `"\w{3}" "abcdef" 3`? although, I can't imagine why it would fail given other tests we have.

Reply:
I tried to keep these tests pretty minimal, given that we know the functionality is using the same entry point as Replace/Split/etc.  So the tests are trying to validate the count produced is correct effectively assuming the engine is doing the right thing, rather than actually trying to validate the engine (which would require a lot more tests).

Reply:
yup